crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -880,6 +881,34 @@ where
 
         self.current_match_case.as_mut().unwrap().index += 1;
     }
+
+    fn visit_except_handler(&mut self, except_handler: &'ast ast::ExceptHandler) {
+        let ast::ExceptHandler::ExceptHandler(ast::ExceptHandlerExceptHandler {
+            name,
+            type_,
+            body,
+            range: _,
+        }) = except_handler;
+        if let Some(type_) = type_ {
+            if let Some(symbol_name) = name {
+                self.add_standalone_expression(type_);
+                self.visit_expr(type_);
+                debug_assert!(self.current_assignment.is_none());
+                let symbol =
+                    self.add_or_update_symbol(symbol_name.id.clone(), SymbolFlags::IS_DEFINED);
+                self.add_definition(
+                    symbol,
+                    ExceptHandlerDefinitionNodeRef {
+                        symbol_name,
+                        handled_exceptions: type_,
+                    },
```

Comment:
An `ExceptHandler` node only creates a definition if the `type_` and `name` fields of the `ExceptHandler` are both `Some()`. If we stored a reference to the entire except-handler node here rather than storing references to the symbol name and the handled exceptions (respectively an `Identifier` and an `Expr`), we'd have to do unnecessary type narrowing or unwrapping at later stages. It seems to me like it would be much less type-safe and much less ergonomic.

Reply:
We have to do this type narrowing and unwrapping again in inference either way, because we will run into the exception handler in scope visiting and have to a) decide if it is a kind of except handler that creates a definition, and b) unwrap the relevant parts of it again. The only difference is whether we have to do the unwrapping in `infer_except_handler` or in `infer_except_handler_definition`. So there is no gain in either type safety of ergonomics. But there is a small loss in efficiency and ergonomics on this side due to the extra wrapping types needed.

Reply:
So you want me to do something like this? It seems to me like there is clearly a loss of type safety here: I have to add a `.expect()` call in `infer_except_handler_definition()` that's unnecessary in the code as I currently have it:

<details>
<summary>Diff</summary>

```diff
diff --git a/crates/red_knot_python_semantic/src/semantic_index/builder.rs b/crates/red_knot_python_semantic/src/semantic_index/builder.rs
index 668ad4b77..0c00472c3 100644
--- a/crates/red_knot_python_semantic/src/semantic_index/builder.rs
+++ b/crates/red_knot_python_semantic/src/semantic_index/builder.rs
@@ -883,12 +883,13 @@ where
     }
 
     fn visit_except_handler(&mut self, except_handler: &'ast ast::ExceptHandler) {
-        let ast::ExceptHandler::ExceptHandler(ast::ExceptHandlerExceptHandler {
+        let ast::ExceptHandler::ExceptHandler(except_handler) = except_handler;
+        let ast::ExceptHandlerExceptHandler {
             name,
             type_,
             body,
             range: _,
-        }) = except_handler;
+        } = except_handler;
         if let Some(handled_exceptions) = type_ {
             self.visit_expr(handled_exceptions);
             if let Some(symbol_name) = name {
@@ -898,7 +899,7 @@ where
                     symbol,
                     ExceptHandlerDefinitionNodeRef {
                         symbol_name,
-                        handled_exceptions,
+                        except_handler,
                     },
                 );
             }
diff --git a/crates/red_knot_python_semantic/src/semantic_index/definition.rs b/crates/red_knot_python_semantic/src/semantic_index/definition.rs
index cce6de9db..aa05fa468 100644
--- a/crates/red_knot_python_semantic/src/semantic_index/definition.rs
+++ b/crates/red_knot_python_semantic/src/semantic_index/definition.rs
@@ -172,7 +172,7 @@ pub(crate) struct ComprehensionDefinitionNodeRef<'a> {
 #[derive(Copy, Clone, Debug)]
 pub(crate) struct ExceptHandlerDefinitionNodeRef<'a> {
     pub(crate) symbol_name: &'a ast::Identifier,
-    pub(crate) handled_exceptions: &'a ast::Expr,
+    pub(crate) except_handler: &'a ast::ExceptHandlerExceptHandler,
 }
 
 #[derive(Copy, Clone, Debug)]
@@ -262,11 +262,9 @@ impl DefinitionNodeRef<'_> {
                 index,
             }),
             DefinitionNodeRef::ExceptHandler(ExceptHandlerDefinitionNodeRef {
-                handled_exceptions,
+                except_handler,
                 ..
-            }) => DefinitionKind::ExceptHandler(ExceptHandlerDefinitionKind {
-                handled_exceptions: AstNodeRef::new(parsed, handled_exceptions),
-            }),
+            }) => DefinitionKind::ExceptHandler(AstNodeRef::new(parsed.clone(), except_handler)),
         }
     }
 
@@ -322,18 +320,7 @@ pub enum DefinitionKind {
     ParameterWithDefault(AstNodeRef<ast::ParameterWithDefault>),
     WithItem(WithItemDefinitionKind),
     MatchPattern(MatchPatternDefinitionKind),
-    ExceptHandler(ExceptHandlerDefinitionKind),
-}
-
-#[derive(Clone, Debug)]
-pub struct ExceptHandlerDefinitionKind {
-    handled_exceptions: AstNodeRef<ast::Expr>,
-}
-
-impl ExceptHandlerDefinitionKind {
-    pub(crate) fn handled_exceptions(&self) -> &ast::Expr {
-        self.handled_exceptions.node()
-    }
+    ExceptHandler(AstNodeRef<ast::ExceptHandlerExceptHandler>),
 }
 
 #[derive(Clone, Debug)]
diff --git a/crates/red_knot_python_semantic/src/types/infer.rs b/crates/red_knot_python_semantic/src/types/infer.rs
index 3ddf821ae..486e34fc4 100644
--- a/crates/red_knot_python_semantic/src/types/infer.rs
+++ b/crates/red_knot_python_semantic/src/types/infer.rs
@@ -426,10 +426,7 @@ impl<'db> TypeInferenceBuilder<'db> {
                 );
             }
             DefinitionKind::ExceptHandler(except_handler_definition) => {
-                self.infer_except_handler_definition(
-                    except_handler_definition.handled_exceptions(),
-                    definition,
-                );
+                self.infer_except_handler_definition(except_handler_definition, definition);
             }
         }
     }
@@ -805,9 +802,12 @@ impl<'db> TypeInferenceBuilder<'db> {
 
     fn infer_except_handler_definition(
         &mut self,
-        handled_exceptions: &'db ast::Expr,
+        except_handler: &'db ast::ExceptHandlerExceptHandler,
         definition: Definition<'db>,
     ) {
+        let handled_exceptions = except_handler.type_.as_deref().expect(
+            "`type_` should always be `Some()` for an exception handler that creates a definition",
+        );
         let node_ty = self.infer_expression(handled_exceptions);
 
         // TODO: anything that's a consistent subtype of
```

</details>

Reply:
Good point, yeah. I think that's good enough reason that the single node we pass through the definition can be the handled-types expression rather than the entire handler. But I still think we have unnecessary complexity that can be removed. Will make it a different comment, since this thread is on "outdated" code in the PR.